fdbserver/ptxn/TLogServer.actor.cpp

```diff
@@ -1500,30 +1513,285 @@ void stopAllTLogs(Reference<TLogServerData> self, UID newLogId) {
+				ASSERT(!storageTeamData);
+				logData->createStorageTeamData(id, tags, popped);
+				logData->getStorageTeamData(id)->persistentPopped = popped;
+			}
 		}
 	}
+
+	std::sort(logsByVersion.begin(), logsByVersion.end());
+	for (const auto& pair : logsByVersion) {
+		// TLogs that have been fully spilled won't have queue entries read in the loop below.
+		self->popOrder.push_back(pair.second);
+	}
+	logsByVersion.clear();
+
+	state Future<Void> allRemoved = waitForAll(removed);
+	state UID lastId = UID(1, 1); // initialized so it will not compare equal to a default UID
+	state double recoverMemoryLimit = SERVER_KNOBS->TLOG_RECOVER_MEMORY_LIMIT;
+	if (BUGGIFY)
+		recoverMemoryLimit =
+		    std::max<double>(SERVER_KNOBS->BUGGIFY_RECOVER_MEMORY_LIMIT, (double)SERVER_KNOBS->TLOG_SPILL_THRESHOLD);
+
+	try {
+		bool recoveryFinished = wait(self->persistentQueue->initializeRecovery(minimumRecoveryLocation));
+		if (recoveryFinished) {
+			throw end_of_stream();
+		}
+		loop {
+			if (allRemoved.isReady()) {
+				TEST(true); // all tlogs removed during queue recovery
+				throw worker_removed();
+			}
+			choose {
+				when(TLogQueueEntry qe = wait(self->persistentQueue->readNext(self.getPtr()))) {
+					if (qe.id != lastId) {
+						lastId = qe.id;
+						auto it = self->id_data.find(qe.id);
+						if (it != self->id_data.end()) {
+							logData = it->second;
+						} else {
+							logData = Reference<LogGenerationData>();
+						}
+					}
+
+					//TraceEvent("TLogRecoveredQE", self->dbgid).detail("LogId", qe.id).detail("Ver", qe.version).detail("MessageBytes", qe.messages.size()).detail("Tags", qe.tags.size())
+					//	.detail("Tag0", qe.tags.size() ? qe.tags[0].tag : invalidTag).detail("Version",
+					// logData->version.get());
+
+					if (logData) {
+						if (!self->spillOrder.size() || self->spillOrder.back() != qe.id) {
+							self->spillOrder.push_back(qe.id);
+						}
+						logData->knownCommittedVersion =
+						    std::max(logData->knownCommittedVersion, qe.knownCommittedVersion);
+						if (qe.version > logData->version.get()) {
+							for (int i = 0; i < qe.messages.size(); i++) {
+								commitMessages(self, logData, qe.version, qe.messages[i], qe.storageTeams[i]);
+							}
+							logData->version.set(qe.version);
+							logData->queueCommittedVersion.set(qe.version);
+
+							while (self->bytesInput - self->bytesDurable >= recoverMemoryLimit) {
+								TEST(true); // Flush excess data during TLog queue recovery
+								TraceEvent("FlushLargeQueueDuringRecovery", self->dbgid)
+								    .detail("LogId", logData->logId)
+								    .detail("BytesInput", self->bytesInput)
+								    .detail("BytesDurable", self->bytesDurable)
+								    .detail("Version", logData->version.get())
+								    .detail("PVer", logData->persistentDataVersion);
+
+								// choose {
+								// 	when(wait(updateStorage(self))) {}
+								// 	when(wait(allRemoved)) { throw worker_removed(); }
+								// }
```

Comment:
Why comment these out? Maybe add TODOs here?

Reply:
uncommented, but in order to do this, I have to re-order the whole file (updateStorage is declared after this reference, and updateStorage itself has other dependent methods, etc). If you prefer, I can also make it work with forward declaration(I think?). @jzhou77

Reply:
Actually let me use forward declaration in this PR to make reviewer see it clearly, will have follow up PR to re-order and remove the forward declaration if necessary.